core/app/models/spree/shipment.rb

```diff
@@ -430,5 +508,20 @@ def ensure_can_destroy
         throw :abort
       end
     end
+
+    def store_state_change(previous_state, new_state)
+      state_changes.create!(
+        previous_state: previous_state,
+        next_state:     new_state,
+        name:           'shipment'
+      )
+    end
+
+    def change_state!(new_state)
+      previous_state = state
+      return if new_state == previous_state
```

Comment:
We may can make use of `ActiveModel::Dirty` here?

http://api.rubyonrails.org/classes/ActiveModel/Dirty.html